services/121-service/src/registration/registrations.service.spec.ts

```diff
@@ -0,0 +1,414 @@
+        // Provide proper mocks for all dependencies
+        {
+          provide: LookupService,
+          useValue: {
+            lookupAndCorrect: jest.fn(),
+          },
+        },
+        {
+          provide: MessageQueuesService,
+          useValue: {
+            addMessageJob: jest.fn(),
+          },
+        },
+        {
+          provide: InclusionScoreService,
+          useValue: {
+            calculateInclusionScore: jest.fn(),
+            calculatePaymentAmountMultiplier: jest.fn(),
+          },
+        },
+        {
+          provide: RegistrationsImportService,
+          useValue: {
+            importRegistrationsFromCsv: jest.fn(),
+            importRegistrations: jest.fn(),
+          },
+        },
+        {
+          provide: RegistrationDataService,
+          useValue: {
+            saveData: jest.fn(),
+            deleteProgramRegistrationAttributeData: jest.fn(),
+          },
+        },
+        {
+          provide: IntersolveVisaService,
+          useValue: {
+            hasIntersolveCustomer: jest.fn(),
+            retrieveAndUpdateWallet: jest.fn(),
+          },
+        },
+        {
+          provide: RegistrationsPaginationService,
+          useValue: {
+            getPaginate: jest.fn(),
+            getRegistrationViewsByReferenceIds: jest.fn(),
+          },
+        },
+        {
+          provide: UserService,
+          useValue: {
+            getProgramScopeIdsUserHasPermission: jest.fn(),
+          },
+        },
+        {
+          provide: RegistrationUtilsService,
+          useValue: {
+            save: jest.fn(),
+          },
+        },
+        {
+          provide: ProgramFinancialServiceProviderConfigurationRepository,
+          useValue: {
+            findOne: jest.fn(),
+            getPropertiesByNamesOrThrow: jest.fn(),
+          },
+        },
+        {
+          provide: RegistrationDataScopedRepository,
+          useValue: {
+            getRegistrationDataArrayByName: jest.fn(),
+            createQueryBuilder: jest.fn().mockReturnValue({
+              leftJoinAndSelect: jest.fn().mockReturnThis(),
+              andWhere: jest.fn().mockReturnThis(),
+              getMany: jest.fn(),
+            }),
+          },
+        },
+        {
+          provide: RegistrationsInputValidator,
+          useValue: {
+            validateAndCleanInput: jest.fn(),
+          },
+        },
+      ],
+    }).compile();
+
+    service = module.get<RegistrationsService>(RegistrationsService);
+    registrationScopedRepository = module.get<RegistrationScopedRepository>(
+      RegistrationScopedRepository,
+    );
+    distinctRegistrationPairRepository =
+      module.get<DistinctRegistrationPairRepository>(
+        DistinctRegistrationPairRepository,
+      );
+    eventsService = module.get<EventsService>(EventsService);
+
+    // Spy on the private method to test how it's called
+    jest
+      .spyOn(service as any, 'createRegistrationDistinctPair')
```

Comment:
Co-pilot came up with a nice trick to access private methods in unit tests

Reply:
I refactored this to use the public methods of repositories instead. It was not too much of a hassle, so that's nice